pkg/services/pluginsintegration/angulardetectorsprovider/dynamic.go

```diff
@@ -0,0 +1,302 @@
+package angulardetectorsprovider
+
+import (
+	"context"
+	"encoding/json"
+	"errors"
+	"fmt"
+	"net"
+	"net/http"
+	"net/url"
+	"sync"
+	"time"
+
+	"github.com/grafana/grafana/pkg/plugins/config"
+	"github.com/grafana/grafana/pkg/plugins/log"
+	"github.com/grafana/grafana/pkg/plugins/manager/loader/angular/angulardetector"
+	"github.com/grafana/grafana/pkg/services/pluginsintegration/angularpatternsstore"
+)
+
+const (
+	defaultBackgroundJobInterval = time.Hour * 1
+
+	cacheRestoreTimeout = time.Second * 10
+	gcomFetchTimeout    = time.Minute * 1
+)
+
+// backgroundJob implements the dynamic angular detectors provider job that is periodically executed in the background.
+type backgroundJob interface {
+	// runBackgroundJob updates the dynamic angular detectors from a source.
+	runBackgroundJob(ctx context.Context)
+}
+
+var (
+	_ backgroundJob  = &Dynamic{}
+	_ DynamicUpdater = &Dynamic{}
+)
+
+// Dynamic is an angulardetector.DetectorsProvider that calls GCOM to get Angular detection patterns,
+// converts them to detectors and caches them for all future calls.
+// It also provides a background service that will periodically refresh the patterns from GCOM.
+type Dynamic struct {
+	log log.Logger
+
+	httpClient http.Client
+	baseURL    string
+
+	// store is the underlying angular patterns store used as a cache.
+	store angularpatternsstore.Service
+
+	// detectors contains the cached angular detectors, which are created from the remote angular patterns.
+	// Use setDetectors and ProvideDetectors to write/read this value.
+	detectors []angulardetector.AngularDetector
+
+	// mux is the mutex used to read/write the cached detectors in a concurrency-safe way.
+	mux sync.RWMutex
+
+	// initialRestore is a channel that can be used to await the initial restore to be done.
+	// It is closed when the initial restore from cache is completed.
+	// To wait for the initial restore to be done, simply read a value from this channel.
+	initialRestore chan struct{}
+
+	// backgroundJob is the implementation of the background job. This is periodically invoked by Run().
+	backgroundJob backgroundJob
```

Comment:
Yes, it's for tests and I'm not a big fan of this either. However, the PR went to different iterations and with the current one I think it can be achieved like you suggested (as long as we lock the mutex before fetching patterns from GCOM and release it after we update the store etc). I will take a look :+1: